lib/fastlane/plugin/firebase_app_distribution/actions/firebase_app_distribution_action.rb

```diff
@@ -200,6 +205,72 @@ def self.release_notes(params)
         release_notes_param || Actions.lane_context[SharedValues::FL_CHANGELOG]
       end
 
+      def self.poll_release(client, operation, binary_type)
+        operation = client.get_project_app_release_operation(operation.name)
+        MAX_POLLING_RETRIES.times do
+          if operation.done && operation.response && operation.response['release']
+            release = extract_release(operation)
+            result = operation.response['result']
+            if result == 'RELEASE_UPDATED'
+              UI.success("✅ Uploaded #{binary_type} successfully; updated provisioning profile of existing release #{release_version(release)}.")
+              break
+            elsif result == 'RELEASE_UNMODIFIED'
+              UI.success("✅ The same #{binary_type} was found in release #{release_version(release)} with no changes, skipping.")
+              break
+            else
+              UI.success("✅ Uploaded #{binary_type} successfully and created release #{release_version(release)}.")
+            end
+            break
+          elsif !operation.done
+            sleep(POLLING_INTERVAL_SECONDS)
+            operation = client.get_project_app_release_operation(operation.name)
+          else
+            if operation.error && operation.error.message
+              UI.user_error!("#{ErrorMessage.upload_binary_error(binary_type)}: #{operation.error.message}")
+            else
+              UI.user_error!(ErrorMessage.upload_binary_error(binary_type))
+            end
+          end
+        end
+        extract_release(operation)
+      end
+
+      def self.upload_binary(app_name, binary_path, client, timeout)
+        options = Google::Apis::RequestOptions.new
+        options.max_elapsed_time = timeout
+        options.header = {
+          'Content-Type' => 'application/octet-stream',
+          'X-Goog-Upload-File-Name' => File.basename(binary_path),
+          'X-Goog-Upload-Protocol' => 'raw'
+        }
+        # For some reason calling the client.upload_medium returns nil when
```

Comment:
Ugh. Is the return value the only problem, i.e. does the upload itself work?

Would be really nice to figure this out. I miss @jladieu! We also have other Ruby experts: Maybe @jakeouellette can help? Or @i2amsam?

Or we ask the Googlers who wrote the Ruby client: @bajajneha27 @dazuma @sqrrrl

Reply:
How dare you question my ruby skills! 😛 

I don't think this is necessarily a ruby problem but more of a google api client issues. The upload actually does work, and you can see the correct response in the debug log, but something weird must be happening when the response is being deserialized, which is causing the response to come back as _nil_. We need the operation name from the response to poll the status of the upload so we're screwed if we don't get the information from the response.

FWIW I don't think this is that bad of a workaround. We would have to set these custom headers anyway since I _believe_ we only support raw uploads and need to include the file name to determine if it's an APK or an AAB.  This is still better than what we were doing in the custom API client.

Reply:
Since the upload works, I agree that it's probably a bug in the API client code.  My idea was to debug the API client code to find out what's wrong/weird and get that fixed. It should not return `nil`.

Our server _should_ support all types of uploads, since it's using Scotty.

As for the custom headers, I looked at the [code](https://github.com/googleapis/google-api-ruby-client/blob/main/google-apis-core/lib/google/apis/core/upload.rb) and unfortunately `X-Goog-Upload-File-Name` is not set. I guess we could/should fix that. The code does set the `Content-Type` header based on the file extension, though. So we could update our server-side code to fall back to that if there's no filename.

I guess I'm fine with leaving a `TODO(kbolay): ...` for now.

Reply:
Done.